### PR TITLE
Purged provenance-specific constructs from the AST.

### LINKED ISSUE #1705: Refactor/Modularise AST-to-RAM translator
The translator is currently a huge mess that's been left aside in favour of improving the AST and RAM representations. Now that the AST and RAM structure has been mostly fixed up, the translator should be cleaned up and modularised into a more maintainable and clear form.

Work has already begun, with PRs #1665 and #1676 moving out the clause translator, value index, and location classes into separate independent files. Similar work should be done for value, constraint, and relation translation. The current goal is to break up the functionality/role of the translator into smaller, more manageable pieces.

### PR DESCRIPTION
**Pull request by azreika**
Completely eliminated provenance elements from the AST, including:
* The provenance transformer (adding attributes to all relations)
* Info clause/relation creation
* Built-in eqrel expansion (now happens as a separate conditional AST transformer)
* All notions of auxiliary arity prior to the translator

This marks the last major PR for the redesign of the AST->RAM translator! Provenance is now purely a translation strategy.

Some final minor work still needs to be done, listed in the TODOs, but the major structure and design choices have been done. Changes to provenance should now be done via `ast2ram/provenance/` changes.

This PR resolves issue #1709. Since the main translator refactoring is now complete, this PR also resolves #1705.

### CODE REVIEW

**Comment by TomasPuverle** [comment]
@azreika, Would you please not merge until after my PRs?  I would have to rebase 5 PRs.

**Comment by azreika** [comment]
> @azreika, Would you please not merge until after my PRs? I would have to rebase 5 PRs.

No problem.

**Comment by TomasPuverle** [comment]
Thank you, much appreciated.

File: src/ast/transform/ExpandEqrels.cpp
Comment (TomasPuverle) [inline src/ast/transform/ExpandEqrels.cpp:77]: This will no longer work after merging my code change.  Head is always required.  I suggest 

```suggestion
        auto transitiveClause = mk<Clause>(relName);
        auto transitiveClauseHead = transitiveClause->getHead();
```

File: src/ast/transform/ExpandEqrels.h
Comment (TomasPuverle) [inline src/ast/transform/ExpandEqrels.h:31]: FYI: These are now called `cloneImpl` after my PRs.

File: src/ast/transform/IOAttributes.h
Comment (TomasPuverle) [inline src/ast/transform/IOAttributes.h:90]: DRY?  `auto` perhaps?

File: src/ast2ram/provenance/SubproofGenerator.cpp
Comment (TomasPuverle) [inline src/ast2ram/provenance/SubproofGenerator.cpp:178]: Please use `as<>`.  One of my PRs went though the whole codebase and replaced `dynamic_cast`.  It'll also make my future work easier ;), and it's shorter to type.

File: src/ast2ram/provenance/SubproofGenerator.cpp
Comment (TomasPuverle) [inline src/ast2ram/provenance/SubproofGenerator.cpp:195]: `as<>`

File: src/include/souffle/io/SerialisationStream.h
Comment (TomasPuverle) [inline src/include/souffle/io/SerialisationStream.h:86]: `relTypes` is a `std::vector<>`; can use a generic for here.

**Comment by TomasPuverle** [comment]
Since I'm also writing the Python module, what does this refactor mean for `souffle::Relation::getAuxiliaryArity()` and `souffle::Relation::getPrimaryArity()`?  Are these gone/should they be removed?

File: src/ast/transform/ExpandEqrels.cpp
Comment (azreika) [reply src/ast/transform/ExpandEqrels.cpp:77]: No problem, I will make these changes. Will lead to better codeflow, I think.

File: src/ast/transform/IOAttributes.h
Comment (azreika) [reply src/ast/transform/IOAttributes.h:90]: I'm keeping it consistent with the rest of the code in the file for now - I just removed the auxiliary arity field. A future PR might look into fixing up this code quality issue in the file.

**Comment by azreika** [comment]
> Since I'm also writing the Python module, what does this refactor mean for `souffle::Relation::getAuxiliaryArity()` and `souffle::Relation::getPrimaryArity()`? Are these gone/should they be removed?

Auxiliary arity is no longer a thing in the AST, but remains relevant on the RAM level. I'm not sure what the module operates on, but if the relation here is referring to an AST relation (i.e. at the Datalog/logic level), then it should be removed. If it's operating on the RAM level or afterwards (i.e. a relation produced after translation), then auxiliary arity is still relevant.

**Comment by azreika** [comment]
These three provenance tests are failing consistently on Jenkins: `high_arity`, `eqrel_tests3`, and `explain_float_unsigned`. The output for `eqrel_tests3` on Jenkins seems to indicate incorrect output (for either the proof tree or actual result). Not sure what the problem is, because I can't seem to reproduce the problem locally (Ubuntu 20.04 LTS) or on a separate remote server (Fedora 32). Still looking into it.

**Comment by azreika** [comment]
OSX tests show an assertion failure in `include/souffle/provenance/ExplainProvenanceImpl.h` with rule numbers:

```
assert(info.find(std::make_pair(relName, ruleNum)) != info.end() && "invalid rule for tuple");
```

Could be an issue with inconsistent rule numbering, or mishandling of information being passed. The file should be cleaned up anyway, so will make a pass through.

**Comment by azreika** [comment]
Problem now seems to be non-determinism with rule-numbering in provenance tests.

**Review by b-scholz — APPROVED** [review]

**Pull request merged**
